keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlResourceRepository.kt

```diff
@@ -183,90 +191,83 @@ open class SqlResourceRepository(
     }
   }
 
-  override fun eventHistory(id: String, limit: Int): List<ResourceEvent> {
+  override fun eventHistory(id: String, limit: Int): List<ResourceHistoryEvent> {
     require(limit > 0) { "limit must be a positive integer" }
+
+    val resource = get(id)
+
     return sqlRetry.withRetry(READ) {
       jooq
         .select(EVENT.JSON)
-        .from(EVENT, RESOURCE)
-        .where(EVENT.SCOPE.eq(Scope.RESOURCE.name))
-        .and(RESOURCE.ID.eq(id))
-        .and(RESOURCE.UID.eq(EVENT.UID))
+        .from(EVENT)
+        // look for resource events that match the resource...
+        .where(EVENT.SCOPE.eq(Scope.RESOURCE.name)
+          .and(EVENT.UID.eq(resource.uid))
+        )
+        // ...or application events that match the application as they apply to all resources
+        .or(EVENT.SCOPE.eq(Scope.APPLICATION.name)
+          .and(EVENT_JSON_APPLICATION.eq(resource.application))
+        )
```

Comment:
This is the gist of the change to query the right event records from the database, which hopefully is pretty straightforward to read.